tests/jballerina-unit-test/src/test/resources/test-src/record/map_to_record.bal

```diff
@@ -0,0 +1,208 @@
+
+    map<string> mString = {"a": "foo", "b": "bar"};
+    record {} rString = mString;
+    rString["c"] = "baz";
+    assert(<record {}>{"a": "foo", "b": "bar", "c": "baz"}, rString);
+
+    map<json> mJson = {"a": {"foo": "bar"}};
+    record {} rJson = mJson;
+    rJson["b"] = <json>{"baz": "qux"};
+    assert(<record {}>{"a": {"foo": "bar"}, "b": {"baz": "qux"}}, rJson);
+
+    // TODO: disabled due to https://github.com/ballerina-platform/ballerina-lang/issues/22276
+    //map<xml> mXml = {"a": xml `<book>The Lost World</book>`};
+    //record {} rXml = mXml;
+    //rXml["b"] = xml `<book>Sherlock Holmes</book>`;
+    //assert(<record {}>{"a": xml `<book>The Lost World</book>`, "b": xml `<book>Sherlock Holmes</book>`},
+    //       rXml);
+}
+
+function testInherentTypeViolationInInclusiveRecords() {
+    map<decimal> mDecimal = {"a": 1.2, "b": 2.3};
+    record {} rDecimal = mDecimal;
+    rDecimal["c"] = 3.4; // this panics
+}
+
+function testInherentTypeViolationInExclusiveRecords() {
+    map<int> mInt = {"a": 1, "b": 2};
+    record {| (int|string)...; |} r = mInt;
+    r["c"] = "foo"; // this panics
+}
+
+function testRecordsWithOptionalFields() {
+    map<string> m = {"a":"aaa", "b":"bbb"};
+
+    record {|
+        string a?;
+        string c?;
+        string...;
+    |} r = m;
+
+    assert("aaa", r?.a);
+    assert((), r?.c);
+
+    r.a = "AA";
+    r.c = "CC";
+
+    assert(<record {| string a?; string c?; string...; |}>{"a": "AA", "b": "bbb", "c": "CC"}, r);
+}
+
+type Foo record {|
+    int a;
+    int...;
+|};
+
+type Bar record {|
+    int a;
+|};
+
+function testSubtyping() {
+    map<Bar> m1 = {"bb": {a: 10}};
+
+    record {|
+        Foo...;
+    |} r1 = m1;
+
+    Foo f1 = r1.get("bb");
+
+    assert(<Foo>{a: 10}, f1);
+    assert(<Bar>{a: 10}, f1);
+
+    f1["cc"] = 20; // this panics
+}
+
+type Baz record {|
+    *Bar;
+    int b?;
+    int...;
+|};
+
+type SuperTypeOfAll record {|
+    int...;
+|};
+
+// Bar is a subtype of Foo
+map<Bar> gM = {"bb": {a: 10}, "x": {a: 20}};
+
+record {|
+    SuperTypeOfAll x?;
+    Foo...;
+|} gR = gM;
+
+function testComplexSubtyping() {
+    // Fails since although the literal {a:10} is a valid value, its type is Bar.
+    Baz b1 = <Baz>gR.get("x");
+}
+
+function testComplexSubtyping2() {
+    record {| int c; |} rC = {c: 50};
+
+    // Compile time allowed since rC's type is a subtype of SuperTypeOfAll. But fails at run time since gR is actually
+    // a Bar map value.
+    gR.x = rC;
+}
+
+
+// Util functions
+
+function assert(anydata expected, anydata actual) {
+    if (expected != actual) {
+        typedesc<anydata> expT = typeof expected;
+        typedesc<anydata> actT = typeof actual;
+        string reason = "expected [" + expected.toString() + "] of type [" + expT.toString()
+                            + "], but found [" + actual.toString() + "] of type [" + actT.toString() + "]";
+        panic error(reason);
```

Comment:
Shall we use a simple literal (e.g., `AssertionError`) and use the current reason as the `message` of the detail?